k2/python/k2/mutual_information.py

```diff
@@ -0,0 +1,298 @@
+        where N is the number of terms that the sum over t' included, which
+        might include some or all of the other sequences as well as this one.
+
+        Note: we don't require px and py to be contiguous, but the
+        code assumes for optimization purposes that the T axis has
+        stride 1.
+
+      py:
+        A torch.Tensor of the same dtype as px, with shape [B][S+1][T],
+        representing
+
+          py[b][s][t] =  log [ p(y_t | x_{0..s-1}, y_{0..t-1}) / p(y_t) ]
+
+        This function does not treat x and y differently; the only difference
+        is that for optimization purposes we assume the last axis
+        (the t axis) has stride of 1; this is true if px and py are
+        contiguous.
+
+      boundary:
+        If supplied, a torch.LongTensor of shape [B][4], where each
+        row contains [s_begin, t_begin, s_end, t_end],
+        with 0 <= s_begin <= s_end < S and 0 <= t_begin <= t_end < T
+        (this implies that empty sequences are allowed).
+        If not supplied, the values [0, 0, S, T] will be assumed.
+        These are the beginning and one-past-the-last positions in the x and
+        y sequences respectively, and can be used if not all sequences are
+        of the same length.
+
+      return_grad:
+        Whether to return grads of px and py, this grad standing for the
+        occupation probability is the output of the backward with a
+        `fake gradient` input (all ones)  This is useful to implement the
+        pruned version of rnnt loss.
+
+    Returns:
+        Returns a torch.Tensor of shape [B], containing the log of the mutual
+        information between the b'th pair of sequences.  This is defined by
+        the following recursion on p[b,s,t] (where p is of shape [B,S+1,T+1]),
+        representing a mutual information between sub-sequences of lengths s
+        and t:
+
+             p[b,0,0] = 0.0
+             p[b,s,t] = log_add(p[b,s-1,t] + px[b,s-1,t],
+                                p[b,s,t-1] + py[b,s,t-1])
+                       (if s > 0 or t > 0)
+
+        where we handle edge cases by treating quantities with negative indexes
+        as -infinity.  The extension to cases where the boundaries are specified
+        should be obvious; it just works on shorter sequences with offsets into
+        px and py.
+    """
+    assert px.ndim == 3
+    B, S, T1 = px.shape
+    T = T1 - 1
+    assert py.shape == (B, S + 1, T)
+    assert px.dtype == py.dtype
+    (B, S, T) = px.shape
+    if boundary is not None:
+        assert boundary.dtype == torch.int64
+        assert boundary.shape == (B, 4)
+        for [s_begin, t_begin, s_end, t_end] in boundary.to('cpu').tolist():
+            assert 0 <= s_begin <= s_end <= S
+            assert 0 <= t_begin <= t_end <= T
+    # The following assertions are for efficiency
+    assert px.stride()[-1] == 1
+    assert py.stride()[-1] == 1
+    m, px_grad, py_grad = MutualInformationRecursionFunction.apply(
```

Comment:
If `return_grad` is False, is it possible to not return `px_grad` and `py_grad` to save memory?

Reply:
... that is a valid point.  In some circumstances we might not even need the backprop.  I'm pretty sure there was a time when it allowed to delay the backward part of the computation, it must have been simplified since then.
The memory requirements are not very huge because it's only one element per (s,t) position, not C elements, but it's still nonzero.